validator/sawtooth_validator/gossip/gossip.py

```diff
@@ -137,49 +240,278 @@ def broadcast(self, gossip_message, message_type, exclude=None):
+
+        if self._peering_mode == 'dynamic':
+            super().start()
+
+    def run(self):
+        while not self._stopped:
+            peers = self._gossip.get_peers()
+            if len(peers) < self._min_peers:
+                LOGGER.debug("Below minimum peer threshold. "
+                             "Doing topology search.")
+
+                self._reset_candidate_peer_endpoints()
+                self._refresh_peer_list(peers)
+
+                peers = self._gossip.get_peers()
+
+                self._get_peers_of_peers(peers)
+                self._get_peers_of_endpoints(peers,
+                                             self._initial_join_endpoints)
+
+                # Wait for GOSSIP_GET_PEER_RESPONSE messages to arrive
+                time.sleep(self._response_duration)
+
+                peered_endpoints = list(peers.values())
+
+                with self._condition:
+                    unpeered_candidates = list(
+                        set(self._candidate_peer_endpoints) -
+                        set(peered_endpoints) -
+                        set([self._public_uri]))
+
+                LOGGER.debug("Number of peers: %s",
+                             len(peers))
+                LOGGER.debug("Peers are: %s",
+                             list(peers.values()))
+                LOGGER.debug("Unpeered candidates are: %s",
+                             unpeered_candidates)
+
+                if unpeered_candidates:
+                    self._attempt_to_peer_with_endpoint(
+                        random.choice(unpeered_candidates))
+
+            time.sleep(self._check_frequency)
+
+    def stop(self):
+        self._stopped = True
+
+    def add_candidate_peer_endpoints(self, peer_endpoints):
+        """Adds candidate endpoints to the list of endpoints to
+        attempt to peer with.
+
+        Args:
+            peer_endpoints ([str]): A list of public uri's which the
+                validator can attempt to peer with.
+        """
+        with self._condition:
+            for endpoint in peer_endpoints:
+                if endpoint not in self._candidate_peer_endpoints:
+                    self._candidate_peer_endpoints.append(endpoint)
+
+    def _refresh_peer_list(self, peers):
+        for conn_id in peers:
+            try:
+                self._network.get_connection_id_by_endpoint(
+                    peers[conn_id])
+            except KeyError:
+                LOGGER.debug("removing peer %s because "
+                             "connection went away",
+                             peers[conn_id])
+
+                self._gossip.unregister_peer(conn_id)
+
+    def _get_peers_of_peers(self, peers):
+        get_peers_request = GetPeersRequest()
+
+        for conn_id in peers:
+            self._network.send(
+                validator_pb2.Message.GOSSIP_GET_PEERS_REQUEST,
+                get_peers_request.SerializeToString(),
+                conn_id)
```

Comment:
Per above, I'd like to maintain the pattern of msg->ack(ok | error), with separate dispatch action on the response types for now. I think this may support other uses that we haven't thought of yet.